author	Frédéric Dalleau <frederic.dalleau@linux.intel.co
commit	46b3a3d2d00bf70bc57ef0c9ad5542a2271e3350
	Tue, 27 Mar 2012 11:26:25 +0000 (14:26 +0300)

sdp: Fix sdp_get_profile_descs for Mac Os X Lion

Mac OS X 10.7.3 publishes the following SDP record for HFP Gateway.
      record #2
          aid 0x0000 (SrvRecHndl)
             uint 0x10003
          aid 0x0001 (SrvClassIDList)
             < uuid-16 0x111f (Handsfree AG) uuid-16 0x1203 (Audio) >
          aid 0x0004 (ProtocolDescList)
             < < uuid-16 0x0100 (L2CAP) > <
             uuid-16 0x0003 (RFCOMM) uint 0x2 > >
          aid 0x0005 (BrwGrpList)
             < uuid-16 0x1002 (PubBrwsGrp) >
          aid 0x0006 (LangBaseAttrIDList)
             < uint 0x656e uint 0x6a uint 0x100 >
          aid 0x0009 (BTProfileDescList)
             < uuid-16 0x111e (Handsfree) uint 0x105 >
          aid 0x0100 (SrvName)
             str "Hands Free Audio Gateway"
          aid 0x0301 (SuppDataStoresList)
             uint 0x0
          aid 0x0311 (SuppFeatures)
             uint 0x0

BlueZ do not read correctly the version number of the profile.
The patch fixes this.

From my understanding, the SDP profile descriptor is a list. The for
loop in sdp_get_profile_desc iterates that list.  Until now, bluez
expected a collection of either uuid or lists with a uuid and a
version number. This can be represented as follow:
list=(uuid, uuid, (uuid, version))

The patch fixes the situation where the profile descriptor list
contains a uuid AND a version number both on the root level and not
enclosed in a sublist. If the next item after uuid is uint16, then
this is the version number. This can be represented as follow:
list=(uuid, uuid, version, uuid, (uuid, version))

Note that core v4.0 spec. states that the Profile Descriptor
list is a data sequence of profile descriptors. Each profile descriptor
is a data sequence whose first element is a UUID and second element is
a version number. This can be represented as follow:
list=((uuid, version), (uuid, version))

diff -cr bluez-4.99.pan.orig/lib/sdp.c bluez-4.99.pan/lib/sdp.c
*** bluez-4.99.pan.orig/lib/sdp.c	Fri Oct 12 07:52:02 2012
--- bluez-4.99.pan/lib/sdp.c	Fri Oct 12 08:17:28 2012
***************
*** 2064,2070 ****
--- 2064,2075 ----
  		uint16_t version = 0x100;
  
  		if (SDP_IS_UUID(seq->dtd)) {
+ 			sdp_data_t *next = seq->next;
  			uuid = &seq->val.uuid;
+ 			if (next && next->dtd == SDP_UINT16) {
+ 				version = next->val.uint16;
+ 				seq = next;
+ 			}
  		} else if (SDP_IS_SEQ(seq->dtd)) {
  			sdp_data_t *puuid = seq->val.dataseq;
  			sdp_data_t *pVnum = seq->val.dataseq->next;
